File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
    if (n == 0) return 0;
    if (n == 1) return nums[0];

    int[] dp1 = new int[n + 1];
    int[] dp2 = new int[n + 1];

    dp1[0] = 0;
    dp1[1] = nums[0]; // include first
    dp2[0] = 0;
    dp2[1] = 0;       // exclude first

    // Case 1: include first, exclude last
    for (int i = 2; i <= n - 1; i++) {
        dp1[i] = Math.max(dp1[i - 1], dp1[i - 2] + nums[i - 1]);
    }

    // Case 2: exclude first, include last
    for (int i = 2; i <= n; i++) {
        dp2[i] = Math.max(dp2[i - 1], dp2[i - 2] + nums[i - 1]);
    }

    return Math.max(dp1[n - 1], dp2[n]);
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        ln = len(nums)
        if ln == 0:
            return 0
        if ln == 1:
            return nums[0]

        def robCircle(arr: List[int]) -> int:

            if len(arr) == 1:
                return arr[0]
            prev2 = arr[0]
            prev = max(arr[0], arr[1])

            for i in range(2, len(arr)):
                take = arr[i] + prev2
                not_take = 0 + prev
                curi = max(take, not_take)
                prev2 = prev
                prev = curi
            return prev

        return max(robCircle(nums[1:]), robCircle(nums[:-1]))
